py/fetcher/main.py

```diff
@@ -0,0 +1,160 @@
+import certifi
+import logging
+
+from datadog_api_client import ApiClient, Configuration
+from datadog_api_client.v1.api.metrics_api import MetricsApi
+
+from usageaccountant import UsageAccumulator, UsageUnit
+
+
+logger = logging.getLogger("fetcher")
+
+# TODO remove this before prod
+# logging.basicConfig(
+#     level=logging.INFO,
+#     format="%(asctime)s [%(levelname)s] %(message)s",
+#     handlers=[
+#         logging.FileHandler("debug.log"),
+#         logging.StreamHandler()
+#     ]
+# )
+
+
+def is_valid_query(query):
+    """
+    Validates if query string contains shared_resource_id
+    and app_feature–parameters used in UsageAccountant.record().
+    """
+    return "shared_resource_id" in query and "app_feature" in query
+
+
+def get(query, start_time, end_time):
+    """
+    Fetches timeseries data from Datadog API.
+
+    query (str): Datadog query
+    start_time (int): Start of the time window in Unix epoch format with second-level precision
+    end_time (int): End of the time window in Unix epoch format with second-level precision
+    returns model.metrics_query_response.MetricsQueryResponse
+    """
+    configuration = Configuration(ssl_ca_cert=certifi.where())
+    api_client = ApiClient(configuration)
+    api_instance = MetricsApi(api_client)
+    response = api_instance.query_metrics(
+        _from=start_time,
+        to=end_time,
+        query=query
+    )
+
+    return response
+
+
+def is_valid_response(response):
+    """
+    Validates response object for the following,
+    1. presence of non-empty series list
+    2. presence of a unit in the first element of series list
+    3. presence of a unit that's supported by usageaccountant.UsageUnit
+    """
+    error_msg = response.get("error")
+    if error_msg:
+        log_query_details(response, error_msg, logging.ERROR)
+        return False
```

Comment:
On each of these conditions you stop the script. So please raise exceptions instead of logging a an error with no context:
- If you raise and exception, the script will not exit with 0 exit code, and both k8s and any other shell absed system will interpret it as an error. Which means you can have alerts on it.
- The exception is going to be logged with the stack trace, thus more context
- You can plug an error tracker like sentry